### ISSUE TITLE
Sharpness-Aware Minimization for Efficiently Improving Generalization

### ISSUE THREAD

**Issue by reyllama** [reyllama/paper-reviews#30]
![image](https://user-images.githubusercontent.com/46308407/105318637-17473080-5c07-11eb-8a99-00c0dace0634.png)

- Sharp vs wide minimum (left: ordinary, right: SAM)

![image](https://user-images.githubusercontent.com/46308407/105318811-52e1fa80-5c07-11eb-8a2c-2b87a5ee52f1.png)

- Benefit of reaching at wide minimum
- Above two figures explain pretty much **the essence** of SAM. 

# Key Idea

- Over-parameterized models (most currently used deep learning models) typically yield continuous manifolds of optimal *w*.
- Some optima yields good generalization performance while others might not.
- Wide optimum, having low curvature and generally flat landscape in the neighborhood, tends to be more robust to gap between empirical loss and population loss(test loss).
- Propose simple approximations to optimize SAM loss, loss combining training loss with sharpness loss.
- When applied to current SOTA in image classifications, give significant generalization performance boost, achieving SOTA in numerous fields.

# Previous Approaches

- Some works discussed the property of over-parameterized regime (how global optimum can be achieved under certain conditions even without convexity)
- Others shown implicit bias of Gradient Descent on miniature (linear) model regime, arguing that Gradient Descent leads to max margin classifier with (theoretically) best generalization performance.
- Regularization on **sharpness** have been discussed several times, but have failed to present a scalable implementation method, as this work has. (one additional gradient calculation with reasonable approximation)

# Formulation

- We are trying to minimize population loss (L_D(w)) by minimizing the empirical sample loss (L_S(w)).
- In over-parameterized regime, there's no guarantee that good training loss ensures a model with good generalization performance.
- We can possibly bound the population loss with some combination of training loss and model complexity, as shown below.

![image](https://user-images.githubusercontent.com/46308407/105320046-0bf50480-5c09-11eb-90a7-973cfb5df7d9.png)

![image](https://user-images.githubusercontent.com/46308407/105320103-1f07d480-5c09-11eb-9127-fa69c0f472a3.png)

![image](https://user-images.githubusercontent.com/46308407/105320144-2a5b0000-5c09-11eb-8f8d-576bcaa7f111.png)

- \rho is the single hyperparameter greater thanor equal to 0. Typically set to 0.05 or 0.1 in experiments. We need to find epsilon that maximizes the SAM loss.
- Use Taylor Expansion at 0.

![image](https://user-images.githubusercontent.com/46308407/105320283-55ddea80-5c09-11eb-89ec-adf4bc058ff5.png)

- Solve this optimization problem for epsilon using dual-norm formulation.

![image](https://user-images.githubusercontent.com/46308407/105320369-6c844180-5c09-11eb-8a3d-399ffc930a7f.png)

- Note that this formula of epsilon includes gradient term of L_S(w).

![image](https://user-images.githubusercontent.com/46308407/105320491-93db0e80-5c09-11eb-9731-5615dc75fc8f.png)

- Hence the second term, differentiating epsilon w.r.t. w implies Hessian compuation, which is expensive.
- For the ease of computation, drop the second term on RHS and approximate.

![image](https://user-images.githubusercontent.com/46308407/105320636-c38a1680-5c09-11eb-9b80-18467852a78e.png)

- To compute this gradient, we have to obtain epsilon first, by performing single differentiation on L_S(w) as shown in the dual problem.
- So as a result, optimization on SAM results in 2 gradient steps every time.

# Experiment Results

### Image Classification Training from Scratch

- Note SOTA for CIFAR-10, CIFAR-100

![image](https://user-images.githubusercontent.com/46308407/105321127-66db2b80-5c0a-11eb-82ed-8af00f662e1a.png)

### Image Classification using Transfer Learning (Fine Tuning)

- Note SOTA on CIFAR-10, CIFAR-100 and ImageNet with additional pretraining data.
- ImageNet SOTA is now secured by [meta pseudo labels](https://arxiv.org/abs/2003.10580)

![image](https://user-images.githubusercontent.com/46308407/105321176-7b1f2880-5c0a-11eb-9c38-c31ddbef45d2.png)

### Robustness to Label Noise

- Since SAM leads to a more robust optimum, it can be speculated that this new training regime might help robustness to label noise as well.
- Comparison with highly sophisticated training procedures, specifically devised to combat label noise.

![image](https://user-images.githubusercontent.com/46308407/105321558-f1bc2600-5c0a-11eb-8d4e-1b614054099f.png)

- Generally **on par**.

### Effect on Hessian Spectrum

- If training based on SAM results in wide optimum, it must have low curvature at convergence point. Examine if that's the case.

![image](https://user-images.githubusercontent.com/46308407/105321673-17e1c600-5c0b-11eb-8269-c4a0f7fd53f4.png)

- Hessian Spectrum approximated using Lanczos algorithm by Ghorbani(2019).
- lambda_max / lambda_5 is commonly used as a proxy for sharpness. Note at convergence, the indicator is significantly lower when trained using SAM.

# Comments

- 그동안 CMSC 828W 강의에서 Deep Learning Optimization의 속성에 대해 다양한 측면을 공부했었는데(role of over-parameterization, implicit bias of GD) 모두 이론적인 측면에 치중한 나머지 굉장히 단순한 문제 상황 및 모형(거의 선형..)을 상정한 점이 아쉬웠다. 본 논문에서는 이론적인 부분은 좀 relax한 반면 실제로 사용되고 있는 SOTA 모형들에 실험하여 효과를 입증하였다는 점이 인상적으로 와닿는다.
- 특히 Sharpness를 규제하는 것이 일반화 성능에 도움이 된다는 것이 꽤 직관적으로도 잘 이해되는 부분이어서, 그런 아이디어가 실제로 효과가 좋았다는 점이 반갑다.
- 어떻게보면, 이렇게 training 방법을 새롭게 짠 것이(loss function 변화) 기존의 단순한 SGD에 의한 수렴값보다 좋다는 사실 자체가 앞서 언급한 논문들의 결론이 큰 모델로 일반화되기 어렵다는 것의 반증이기도 한듯. 즉, over-parameterization과 Gradient Descent의 특성 자체가 좋은 optimum으로 안내해준다면, SAM이 이렇게 큰 효과를 가져오면 안되는 것.
- 본 논문에서도 앞서 수학적 분석 부분에, 모집단 손실함수의 upper bound를 training loss+complexity measure로 상정했는데, *Understanding Deep Learning Requires Rethinking Generalization* 논문에서 이야기한 것처럼, 이렇게 단순히 상계를 줄 수 있는 것인지 궁금하다. 예컨대, 같은 모형(고로 같은 ||w||)라고 할지라도 training label을 섞으면 training loss는 여전히 0에 근사하지만 일반화 loss는 치솟을 수 있을듯. 그런 것까지 퉁쳐서 h(||w||) 꼴로 둔 것인지도.
- 하지만 그러나 저러나 이론적인 부분과는 무관하게 실험이 워낙 탄탄하고 성과가 좋아서, 어쨌든 이러한 발상이 지금으로썬 꽤 유효하다고 봐야할 것 같다.
